### ISSUE TITLE
Support for sdag entry method with rdma parameter

### ISSUE THREAD

**Issue by nitbhat** [UIUC-PPL/charm#1553]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/1553*

---
I tried adding rdma functionality to the receiveGhosts method in examples/charm++/load_balancing/stencil3d. The closure structure and the unmarshalling code weren't getting generated correctly as it treats the rdma param as a regular pointer instead of a CkRdmaWrapper.  SDAG support is something we ignored and didn't test in the generic layer patch.  This is more of a feature than a bug.

**Comment by nitbhat** [comment]
*Original date: 2017-05-23 18:47:27*

---
I have finished implementing the SDAG code for rdma entry methods. The usecase for this is the stencil3d example where receiveGhosts is written with ghosts being an rdma parameter.  I have written this example and it works when run without a load balancer. (i.e. without migration)

I am currently working on a migration related bug that I'm seeing when I run the stencil3d example with a load balancer. The bug occurs when the callback is not called in the iteration after the load balancing iteration for the migrated chare. 

I am planning to commit the SDAG support code (xi) and migration example (stencil3d) as two separate commits.

**Comment by PhilMiller** [comment]
*Original date: 2017-05-29 17:19:19*

---
This may be a rather naive question, but why is the receive side code (as suggested by the mention of closure and unmarshalling code) at all affected by the send-side RDMA? Shouldn't the method ultimately receive a `CkMarshallMessage*` as usual?

**Comment by nitbhat** [comment]
*Original date: 2017-06-01 07:36:23*

---
The receive side code changes because we need to follow the same order in unmarshalling as we did in marshalling i.e the numrdmaops is packed first, followed by the  rdmawrappers and then all the other parameters. The unmarshalling code should also follow this. 

Additionally, the closure structure changes based on CMK_ONESIDED_IMPL (and so does its marshalling/unmarshalling code). It has num_rdma_fields and CkRdmawrappers when there is support and regular pointers otherwise. 

Implementation : ~~https://charm.cs.illinois.edu/gerrit/#/c/2572/~~ https://github.com/UIUC-PPL/charm/commit/364594cab57ba5793d0571bf4f706140dc2cbdb8
I have also added a 'rdma' version of stencil3d that has both SDAG and migration. That is included in the patch.
